src/Features/LanguageServer/Protocol/Handler/Completion/CompletionResolveHandler.cs

```diff
@@ -0,0 +1,98 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Completion;
+using Microsoft.CodeAnalysis.LanguageServer.CustomProtocol;
+using Microsoft.VisualStudio.Text.Adornments;
+using Newtonsoft.Json.Linq;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler
+{
+    /// <summary>
+    /// Handle a completion resolve request to add description.
+    /// </summary>
+    [Shared]
+    [ExportLspMethod(LSP.Methods.TextDocumentCompletionResolveName)]
+    internal class CompletionResolveHandler : IRequestHandler<LSP.CompletionItem, LSP.CompletionItem>
+    {
+        public async Task<LSP.CompletionItem> HandleRequestAsync(Solution solution, LSP.CompletionItem completionItem,
+            LSP.ClientCapabilities clientCapabilities, CancellationToken cancellationToken)
+        {
+            CompletionResolveData data;
+            if (completionItem.Data is CompletionResolveData)
+            {
+                data = (CompletionResolveData)completionItem.Data;
+            }
+            else
+            {
+                data = ((JToken)completionItem.Data).ToObject<CompletionResolveData>();
+            }
+
+            var request = data.CompletionParams;
+
+            var document = solution.GetDocumentFromURI(request.TextDocument.Uri);
+            if (document == null)
+            {
+                return completionItem;
+            }
+
+            var position = await document.GetPositionFromLinePositionAsync(ProtocolConversions.PositionToLinePosition(request.Position), cancellationToken).ConfigureAwait(false);
+
+            var completionService = document.Project.LanguageServices.GetService<CompletionService>();
+            var list = await completionService.GetCompletionsAsync(document, position, cancellationToken: cancellationToken).ConfigureAwait(false);
+            if (list == null)
+            {
+                return completionItem;
+            }
+
+            var selectedItem = list.Items.FirstOrDefault(i => i.DisplayText == data.DisplayText);
+            if (selectedItem == null)
+            {
+                return completionItem;
+            }
+
+            var description = await completionService.GetDescriptionAsync(document, selectedItem, cancellationToken).ConfigureAwait(false);
+
+            LSP.CompletionItem resolvedCompletionItem;
+            if (clientCapabilities?.HasVisualStudioLspCapability() == true)
+            {
+                resolvedCompletionItem = CloneVSCompletionItem(completionItem);
```

Comment:
I am not sure why it is cloning the item.

Reply:
It's creating a new VSCompletionItem from a regular completion item.  Since the the contract just specifies a regular completion item as input, it needs to be converted to a VSCompletionItem before it gets returned.